nibabel/processing.py

```diff
@@ -310,3 +311,82 @@ def smooth_image(img,
     return out_class(sm_data, img.affine, img.header)
+
+
+def conform(from_img,
+            out_shape=(256, 256, 256),
+            voxel_size=(1.0, 1.0, 1.0),
+            order=3,
+            cval=0.0,
+            orientation='RAS',
+            out_class=None):
+    """ Resample image to ``out_shape`` with voxels of size ``voxel_size``.
+
+    Using the default arguments, this function is meant to replicate most parts
+    of FreeSurfer's ``mri_convert --conform`` command. Specifically, this
+    function:
+        - Resamples data to ``output_shape``
+        - Resamples voxel sizes to ``voxel_size``
+        - Reorients to RAS (``mri_convert --conform`` reorients to LIA)
+
+    Unlike ``mri_convert --conform``, this command does not:
+        - Transform data to range [0, 255]
+        - Cast to unsigned eight-bit integer
+
+    Parameters
+    ----------
+    from_img : object
+        Object having attributes ``dataobj``, ``affine``, ``header`` and
+        ``shape``. If `out_class` is not None, ``img.__class__`` should be able
+        to construct an image from data, affine and header.
+    out_shape : sequence, optional
+        The shape of the output volume. Default is (256, 256, 256).
+    voxel_size : sequence, optional
+        The size in millimeters of the voxels in the resampled output. Default
+        is 1mm isotropic.
+    order : int, optional
+        The order of the spline interpolation, default is 3.  The order has to
+        be in the range 0-5 (see ``scipy.ndimage.affine_transform``)
+    cval : scalar, optional
+        Value used for points outside the boundaries of the input if
+        ``mode='constant'``. Default is 0.0 (see
+        ``scipy.ndimage.affine_transform``)
+    orientation : str, optional
+        Orientation of output image. Default is "RAS".
+    out_class : None or SpatialImage class, optional
+        Class of output image.  If None, use ``from_img.__class__``.
+
+    Returns
+    -------
+    out_img : object
+        Image of instance specified by `out_class`, containing data output from
+        resampling `from_img` into axes aligned to the output space of
+        ``from_img.affine``
+    """
+    # Only support 3D images. This can be made more general in the future, once tests
+    # are written.
+    required_ndim = 3
+    if from_img.ndim != required_ndim:
+        raise ValueError("Only 3D images are supported.")
+    elif len(out_shape) != required_ndim:
+        raise ValueError("`out_shape` must have {} values".format(required_ndim))
+    elif len(voxel_size) != required_ndim:
+        raise ValueError("`voxel_size` must have {} values".format(required_ndim))
+
+    # Create fake image of the image we want to resample to.
+    hdr = from_img.header_class()
+    hdr.set_data_shape(out_shape)
+    hdr.set_zooms(voxel_size)
+    dst_aff = hdr.get_best_affine()
+    to_img = from_img.__class__(np.empty(out_shape), affine=dst_aff, header=hdr)
```

Comment:
Consider the following image:

```Python
nb.Nifti1Image(np.zeros((256, 256, 256)), affine=np.eye(4))
```

This is already RAS, 1mm isotropic and 256^3 in dimension. However, only the first octant (bounded by `(0, 0, 0), (0, 0, 128), ... ( 128, 128, 128)`) will be shared with `to_img`. When we resample, I assume that we'll get a heavily cropped image.

Assuming I'm correct, it might be better to construct the affine such that the same point is in the center of the image.

Reply:
![ras_256iso](https://user-images.githubusercontent.com/17690870/78819882-ee6deb80-79a4-11ea-932c-8ebe5f6c1a05.png)
![ras_128iso_conformed](https://user-images.githubusercontent.com/17690870/78820243-7522c880-79a5-11ea-8259-040707a2cd23.png)


I have a 256^3 image with 1^3 voxels in RAS, and I ran it through `nib-conform` to change it to 128^3 with 1^3 voxels. It is cropped, but the center of the original is the center of the output. Does this answer what you were getting at?

Reply:
Nope, try the following:

```Python
import nibabel as nb
import numpy as np

img = nb.load(fname)
nb.Nifti1Image(img.dataobj, np.eye(4), img.header).to_filename("corner.nii.gz")
```

This will place the (0, 0, 0) RAS point in the corner of the image. Conform `corner.nii.gz` and you'll see what I'm talking about.

Reply:
OK, did that, and indeed the conformed `corner.nii.gz` is heavily cropped. How can the affine be constructed to maintain the same center?

Reply:
Need to solve for:

```
ras = affine * ijk
affine' * ijk' = affine * ijk
```

Noting that the affine is

```
[rzs t
   0 1]
```

And `t` is the thing you can change.

If you've got time now, go for it, but I can work out the formula this afternoon.

Reply:
Let me take a shot at it this afternoon

Reply:
OK, this is way above level of understanding... sorry @effigies. i'll have to take linear algebra in grad school

Reply:
it looks like this is the freesurfer implementation of the solution we're looking for https://github.com/freesurfer/freesurfer/blob/ea988613231c850785b13632205467613c88d89d/utils/mri.cpp#L415-L523